grr/server/grr_response_server/gui/api_integration_tests/osquery_test.py

```diff
@@ -0,0 +1,96 @@
+# Lint as: python3
+"""Integration tests for the Osquery flow, its API client and API endpoints."""
+from grr_response_server.gui import api_integration_test_lib
+from grr_response_server.flows.general import osquery as osquery_flow
+from grr.test_lib import osquery_test_lib
+from grr.test_lib import flow_test_lib
+from grr.test_lib import test_lib
+from grr.test_lib import action_mocks
+from grr_response_client.client_actions import osquery as osquery_action
+from grr_response_proto.api import osquery_pb2 as api_osquery_pb2
+from grr_api_client import utils
+
+import json
+
+
+class OsqueryResultsExportTest(api_integration_test_lib.ApiIntegrationTest):
+  """Tests exporting of Osquery results using functionality in the API client"""
+
+  def _RunOsqueryExportResults(self, stdout: str) -> utils.BinaryChunkIterator:
+    client_id = self.SetupClient(0)
+
+    with osquery_test_lib.FakeOsqueryiOutput(stdout=stdout, stderr=""):
+      flow_id = flow_test_lib.TestFlowHelper(
+        osquery_flow.OsqueryFlow.__name__,
+        action_mocks.ActionMock(osquery_action.Osquery),
+        client_id=client_id,
+        token=self.token,
+        query="doesn't matter")
+      result_flow = self.api.Client(client_id=client_id).Flow(flow_id)
+      result_flow.WaitUntilDone()
+
+    format_csv = api_osquery_pb2.ApiGetOsqueryResultsArgs.Format.CSV
+    return result_flow.GetOsqueryResults(format_csv)
+
+  def testExportSomeResults(self):
+
+    stdout = """
+    [
+      { "foo": "quux", "bar": "norf", "baz": "thud" },
+      { "foo": "blargh", "bar": "plugh", "baz": "ztesch" }
+    ]
+    """
+    expected_text = "foo,bar,baz\r\nquux,norf,thud\r\nblargh,plugh,ztesch\r\n"
+    expected_bytes = expected_text.encode("utf-8")
+
+    results_iterator = self._RunOsqueryExportResults(stdout)
+
+    csv_bytes = next(results_iterator)
+    self.assertEqual(expected_bytes, csv_bytes)
+
+  def testExportNoRows(self):
+    stdout = """
+    [
+
+    ]
+    """
+
+    results_iterator = self._RunOsqueryExportResults(stdout)
+    csv_bytes = next(results_iterator)
+
+    expected_text = "\r\n"
+    expected_bytes = expected_text.encode("utf-8")
+    self.assertEqual(expected_bytes, csv_bytes)
+
+  def testExportUnicodeCharacters(self):
+    stdout = """
+    [
+      { "🇬 🇷 🇷": "🔝🔝🔝"}
+    ]
+    """
+
+    results_iterator = self._RunOsqueryExportResults(stdout)
+    csv_bytes = next(results_iterator)
+
+    expected_text = "🇬 🇷 🇷\r\n🔝🔝🔝\r\n"
+    expected_bytes = expected_text.encode("utf-8")
+    self.assertEqual(expected_bytes, csv_bytes)
```

Comment:
The assertions would read better and, in case of a failure, the error messages are human-friendly if you decode CSV bytes rather than encoding expectations:

```python
    output = next(results_iterator).decode("utf-8")
    self.assertEqual(output, "🇬 🇷 🇷\r\n🔝🔝🔝\r\n")
```

This remark applies also below.